### PR TITLE
Allowing users to archive applications

### LINKED ISSUE #67: Allow Locker requests to be archived
Sometimes duplicate locker request are put in, or a request for renewal is put in.  These requests never get assigned and are remaining on the assignment page

Need a button to archive them.  Should be able to delete archived requests after some period (2 years)

From April
```
I also need a way to move an application off of the assign locker page. I’m getting people submitting more than one application sometimes because they decided they wanted another floor or thinking that to renew their locker that they needed to submit another application. I have a over a page of applications which do not need to be assigned a locker but I can’t get remove them from the list.
```

### PR DESCRIPTION
**Pull request by rladdusaw**
Resolves #67

### CODE REVIEW

File: spec/requests/locker_applications_spec.rb [hunk 1]
```diff
@@ -358,5 +370,23 @@
...
+    describe 'PUT /toggle_archived' do
+      it 'allows an admin to archive an application' do
+        locker_application = LockerApplication.create! valid_attributes
+        put toggle_archived_locker_application_url(locker_application)
+        locker_application.reload
+        expect(response).to have_http_status(:found)
+        expect(locker_application.archived).to be(true)
+      end
+
+      it 'allows an admin to unachive an application' do
```
Comment (christinach) [inline spec/requests/locker_applications_spec.rb:383]: @rladdusaw thank you! can you please fix the typo in the spec? unachive -> unarchive

**Comment by christinach** [comment]
@rladdusaw I just saw that there are some conflicts. Can you resolve them?

File: app/models/locker_application.rb [hunk 2]
```diff
@@ -58,9 +58,10 @@ def available_lockers_in_area_and_size(building:)
     context.order(:location)
   end
 
-  def self.search(uid:)
-    return where(complete: true) if uid.blank?
-
-    where(complete: true).joins(:user).where("users.uid = '#{uid}'")
+  def self.search(uid:, archived:)
+    results = where(complete: true)
+    results = results.where("locker_applications.archived = '#{archived}'") unless ActiveModel::Type::Boolean.new.cast(archived)
```
Comment (maxkadel) [inline app/models/locker_application.rb:63]: Shouldn't `archived` always already be cast into a boolean at this point? Do we need to re-cast it here?

File: app/models/locker_application.rb [hunk 3]
```diff
@@ -58,9 +58,10 @@ def available_lockers_in_area_and_size(building:)
     context.order(:location)
   end
 
-  def self.search(uid:)
-    return where(complete: true) if uid.blank?
-
-    where(complete: true).joins(:user).where("users.uid = '#{uid}'")
+  def self.search(uid:, archived:)
```
Comment (maxkadel) [inline app/models/locker_application.rb:61]: Right now this method is doing an awful lot of database lookups - is there a way we can streamline this? Maybe something like 

```ruby
def self.search(uid:, archived:)
  # Raise error if archived isn't already a boolean, log and rescue from error and cast archived to a boolean
  return where(complete: true).where(archived: archived) if uid.blank?

  where(archived: archived).where(complete: true).joins(:user).where("users.uid = '#{uid}'")
end
```

You could maybe use this gem to help verify number of database lookups? https://github.com/civiccc/db-query-matchers

File: app/views/locker_applications/_index_view.html.erb [hunk 4]
```diff
@@ -20,6 +20,9 @@
     <grid-item columns="lg-2 md-4 sm-12">
       <%= link_to 'Edit', edit_locker_application_path(locker_application) %>
       <%= link_to 'Assign Locker', assign_locker_application_path(locker_application) %>
+      <% if current_page?(awaiting_assignment_locker_applications_path) %>
```
Comment (maxkadel) [inline app/views/locker_applications/_index_view.html.erb:23]: Is the method `put` needed if it's in the routes file?

File: app/views/locker_applications/awaiting_assignment.html.erb [hunk 5]
```diff
@@ -7,6 +7,10 @@
 
 <h1>Locker Applications</h1>
 
+<card size="full-width">
```
Comment (maxkadel) [inline app/views/locker_applications/awaiting_assignment.html.erb:10]: Is the method `get` needed here?

File: spec/controllers/locker_applications_controller_spec.rb [hunk 6]
```diff
@@ -33,4 +33,34 @@
       expect(flash[:alert]).to eq('Only administrators have access to the everyone\'s Locker Applications!')
     end
   end
+
```
Comment (maxkadel) [inline spec/controllers/locker_applications_controller_spec.rb:36]: Cool test!

File: spec/models/locker_application_spec.rb [hunk 7]
```diff
@@ -132,11 +134,11 @@
     let!(:locker_application3) { FactoryBot.create :locker_application, complete: true }
```
Comment (maxkadel) [inline spec/models/locker_application_spec.rb:134]: Can you add an `archived: true` case? Also, if you think it's possible the method will receive `archived` as a string, test that case as well? Hrm, it also looks like there's no `complete: false` case? Maybe throw one of those in? (sorry, I know I changed that method too)

I'm also confused by the `archived: nil` case - do you expect that to get passed to this method?

**Review by maxkadel — CHANGES_REQUESTED** [review]
A few questions, especially about the `#search` method and tests.

Thanks for working on this! I know edit conflicts are the pits.

File: spec/models/locker_application_spec.rb (hunk 7 above)
Comment (rladdusaw) [reply spec/models/locker_application_spec.rb:134]: The nil case is intended to catch if the page is loaded without query params, but it should be handled before it reaches this method. I think this test might also break if the casting in the search method is removed so this test should be removed.

File: app/models/locker_application.rb (hunk 3 above)
Comment (rladdusaw) [reply app/models/locker_application.rb:61]: One complication is we can't use `where(archived: archived)` when archived is true unless we want to change the functionality from hiding archived or showing all results. I also implemented it so archiving was only possible for applications on the assignment page, this code is used for search for applications on the all applications page. Maybe the best course is to separate these into different methods?

File: app/views/locker_applications/_index_view.html.erb (hunk 4 above)
Comment (rladdusaw) [reply app/views/locker_applications/_index_view.html.erb:23]: Possibly not necessary, retaining for clarity.

File: app/views/locker_applications/awaiting_assignment.html.erb (hunk 5 above)
Comment (rladdusaw) [reply app/views/locker_applications/awaiting_assignment.html.erb:10]: Possibly not necessary, retaining for clarity.

File: spec/controllers/locker_applications_controller_spec.rb (hunk 6 above)
Comment (rladdusaw) [reply spec/controllers/locker_applications_controller_spec.rb:36]: Thanks!

File: app/models/locker_application.rb (hunk 3 above)
Comment (rladdusaw) [reply app/models/locker_application.rb:61]: Refactored to prevent extra database lookups.

File: app/models/locker_application.rb [hunk 8]
```diff
@@ -58,9 +58,10 @@ def available_lockers_in_area_and_size(building:)
     context.order(:location)
   end
 
-  def self.search(uid:)
-    return where(complete: true) if uid.blank?
+  def self.search(uid:, archived:)
+    return joins(:user).where("users.uid = '#{uid}'").where('locker_applications.archived = false').where(complete: true) if uid.present?
```
Comment (sandbergja) [inline app/models/locker_application.rb:62]: Would a single `where` be more readable here?  I could go either way.

```suggestion
    return joins(:user).where(['users.uid = ? and locker_applications.archived = false and complete = true', uid]) if uid.present?
```

**Review by sandbergja — CHANGES_REQUESTED** [review]
Thanks, @rladdusaw !  @leefaisonr and I looked at it today, and just had one concern about an unused `let` in one of the tests.

**Review by sandbergja — APPROVED** [review]
Thanks, @rladdusaw !

**Pull request merged**